specification/ai/OpenAI.Assistants/tools/models.tsp

```diff
@@ -0,0 +1,196 @@
+  @doc("The parameters the functions accepts, described as a JSON Schema object.")
+  parameters: unknown;
+}
+
+//
+// Tool calls -- provided during and after model evaluation to represent needed and past tool use
+//
+
+@discriminator("type")
+@doc("""
+An abstract representation a tool call, issued by the model in evaluation of a configured tool definition, that must
+be fulfilled and have its outputs submitted before the model can continue.
+""")
+@added(ServiceApiVersions.v2023_02_15_preview)
+model ToolCall {
+  @doc("The object type.")
+  type: string;
+
+  @doc("The ID of the tool call. This ID must be referenced when you submit tool outputs.")
+  id: string;
+}
+
+@doc("""
+A tool call to a code interpreter tool, issued by the model in evaluation of a configured code interpreter tool, that
+represents submitted output needed or already fulfilled by the tool for the model to continue.
+""")
+@added(ServiceApiVersions.v2023_02_15_preview)
+model CodeInterpreterToolCall extends ToolCall {
+  @doc("The object type, which is always 'code_interpreter'.")
+  type: "code_interpreter";
+
+  @projectedName("json", "code_interpreter")
+  @doc("The details of the tool call to the code interpreter tool.")
+  codeInterpreter: CodeInterpreterToolCallDetails;
+}
+
+@doc("""
+A tool call to a retrieval tool, issued by the model in evaluation of a configured retrieval tool, that represents
+submitted output needed or already fulfilled by the tool for the model to continue.
+""")
+@added(ServiceApiVersions.v2023_02_15_preview)
+model RetrievalToolCall extends ToolCall {
+  @doc("The object type, which is always 'retrieval'.")
+  type: "retrieval";
+
+  @doc("The key/value pairs produced by the retrieval tool.")
+  retrieval: Record<string>;
+}
+
+@doc("""
+A tool call to a function tool, issued by the model in evaluation of a configured function tool, that represents
+given function inputs and submitted function outputs needed or already fulfilled by the tool for the model to continue.
+""")
+@added(ServiceApiVersions.v2023_02_15_preview)
+model FunctionToolCall extends ToolCall {
+  @doc("The object type, which is always 'function'.")
+  type: "function";
+
+  @doc("The detailed information about the function called by the model.")
+  function: FunctionToolCallDetails;
+}
+
+// Call details: Code Interpreter
+
+@doc("The detailed information about a code interpreter invocation by the model.")
+@added(ServiceApiVersions.v2023_02_15_preview)
+model CodeInterpreterToolCallDetails {
+  @doc("The input provided by the model to the code interpreter tool.")
+  input: string;
+
+  @doc("The outputs produced by the code interpreter tool back to the model in response to the tool call.")
+  outputs: CodeInterpreterToolCallOutput[];
+}
+
+@discriminator("type")
+@doc("An abstract representation of an emitted output from a code interpreter tool.")
+@added(ServiceApiVersions.v2023_02_15_preview)
+model CodeInterpreterToolCallOutput {
```

Comment:
There are many models that include "Tool", "Call", "Output", and a combination of these words. It makes it seems like some models are related to each other when they are not, and vice versa. For example, this `CodeInterpreterToolCallOutput` model in particular is used specifically when inspecting run steps. It is not related to the `ToolOutput` model used when submitting tool outputs. I'm thinking it could be a good idea to rename this model in order to make this more obvious. What do you think about calling it `RunStepDetailsCodeInterpreterOutput`? And then, the two derived models can be called `RunStepDetailsCodeInterpreterLogOutput` and `RunStepDetailsCodeInterpreterImageOutput`?

Reply:
Ah, we also have a model called `RunStepToolCallDetails` related to this scenario. Then, maybe we should call these `RunStepToolCallDetailsCodeInterpreterOutput`, `RunStepToolCallDetailsCodeInterpreterLogOutput`, etc.

I don't like how long the names are, but I think it's more important to be clear.

Reply:
If we end up having a bunch of models that are only used in the run step scenario, we might even consider making the argument that this is an advanced scenario. In .NET, for example, we could consider putting all of these models in a separate namespace to avoid overwhelming the main namespace.

Reply:
As discussed off-thread: fully and enthusiastically agreed on pursuing this architecture line of inquiry ASAP. We won't block initial release on the conceptual refinement dimensions, but it's important for us to do this quickly (and rapidly, in response to usage and feedback).